projects/gloo/pkg/translator/clusters.go

```diff
@@ -291,6 +298,45 @@ func getCircuitBreakers(cfgs ...*v1.CircuitBreakerConfig) *envoy_config_cluster_
 	return nil
 }
 
+// getPreconnectPolicy follows the naming of the rest of functions here
+// it MAY someday deal with inheritance like the other functions
+// it consumes an ordered list of preconnect policies
+// it returns the first non-nil policy
+func getPreconnectPolicy(cfgs ...*v1.PreconnectPolicy) (*envoy_config_cluster_v3.Cluster_PreconnectPolicy, error) {
+
+	// since we dont want strict reliance on envoys current api
+	// but still able to map as closely as possible
+	// if not nil then convert the gloo configurations to envoy
+	for _, curConfig := range cfgs {
+		if curConfig == nil {
+			continue
+		}
+
+		// we have a policy. However we dont respect proto constraints
+		// therefore we need to check here to see if the applied configuration is
+		// SILLY and warn the user.
+		perUpstream := curConfig.GetPerUpstreamPreconnectRatio()
+		predictive := curConfig.GetPredictivePreconnectRatio()
+
+		eString := ""
+		if perUpstream != nil && (perUpstream.GetValue() < 1 || perUpstream.GetValue() > 3) {
+			eString = fmt.Sprintf("perupstream must be between 1 and 3 if set, it was: %v", perUpstream.GetValue())
+		}
+		if predictive != nil && (predictive.GetValue() < 1 || predictive.GetValue() > 3) {
+			eString += fmt.Sprintf("predictive must be between 1 and 3 if set, it was: %v", perUpstream.GetValue())
+		}
+		if eString != "" {
+			return nil, errors.New("invalid preconnect policy: " + eString)
+		}
```

Comment:
As is, if both errors are present there will be no space
Perhaps something like this:
```suggestion
		eStrings := []string{}
		if perUpstream != nil && (perUpstream.GetValue() < 1 || perUpstream.GetValue() > 3) {
			eStrings = append(eStrings, fmt.Sprintf("perupstream must be between 1 and 3 if set, it was: %v", perUpstream.GetValue()))
		}
		if predictive != nil && (predictive.GetValue() < 1 || predictive.GetValue() > 3) {
			eStrings = append(eStrings, fmt.Sprintf("predictive must be between 1 and 3 if set, it was: %v", perUpstream.GetValue()))
		}
		if len(eStrings) > 0 {
			return nil, fmt.Errorf("invalid preconnect policy: %s", strings.Join(eStrings, "; ")
		}
```

(hand written in browser not guaranteed to compile)

Reply:
I like it, have added it